src/tokio/receiver.rs

```diff
@@ -0,0 +1,114 @@
+use std::future::Future;
```

Comment:
This is a good first pass. I might try to implement this with `stream::unwrap` at some point, might be cleaner. 

Don't worry about it for this PR though.

Reply:
I agree, I prefer higher level functions/combinators for stream signaling, creation and transformation, but a side effect free protocol extraction refactoring was my main goal here.